src/modules/sdexec/sdexec.c

```diff
@@ -0,0 +1,1031 @@
+            sdexec_channel_start_output (proc->err);
+        }
+
+    }
+    /* The finished response is sent when wait status is available.
+     * If there was an exec error, "finished" should not be sent.
+     */
+    if (!proc->finished_response_sent) {
+        if (sdexec_unit_has_finished (proc->unit)) {
+            if (flux_respond_pack (h,
+                                   proc->msg,
+                                   "{s:s s:i}",
+                                   "type", "finished",
+                                   "status",
+                                   sdexec_unit_wait_status (proc->unit)) < 0)
+                flux_log_error (h, "error responding to exec request");
+            proc->finished_response_sent = 1;
+        }
+    }
+    /* If the unit reaches active.exited call StopUnit to cause stdout
+     * and stderr to reach eof, and the unit to transition to inactive.dead.
+     */
+    if (sdexec_unit_state (proc->unit) == STATE_ACTIVE
+        && sdexec_unit_substate (proc->unit) == SUBSTATE_EXITED
+        && proc->finished_response_sent) {
+
+        if (!proc->f_stop) {
+            flux_future_t *f2;
+            sdexec_log_debug (h, "stop %s", sdexec_unit_name (proc->unit));
+            if (!(f2 = sdexec_stop_unit (h,
+                                         proc->ctx->rank,
+                                         sdexec_unit_name (proc->unit),
+                                         "fail"))
+                || flux_future_then (f2, -1, stop_continuation, proc) < 0) {
+                flux_log_error (h, "error initiating unit stop");
+                flux_future_destroy (f2);
+                f2 = NULL;
+            }
+            proc->f_stop = f2;
+        }
+    }
+    /* If the unit reaches failed.failed call ResetFailedUnit to cause stdout
+     * and stderr to reach eof, and the unit to transition to inactive.dead.
+     * We can land here for both a child failure and an exec failure.
+     * Start channel output here in case of the latter so it can be finalized.
+     */
+    if (sdexec_unit_state (proc->unit) == STATE_FAILED
+        && sdexec_unit_substate (proc->unit) == SUBSTATE_FAILED) {
+
+        sdexec_channel_start_output (proc->out);
+        sdexec_channel_start_output (proc->err);
+
+        if (!proc->f_stop) {
+            flux_future_t *f2;
+            sdexec_log_debug (h,
+                              "reset-failed %s",
+                              sdexec_unit_name (proc->unit));
+            if (!(f2 = sdexec_reset_failed_unit (h,
+                                                 proc->ctx->rank,
+                                                 sdexec_unit_name (proc->unit)))
+                || flux_future_then (f2, -1, reset_continuation, proc) < 0) {
+                flux_log_error (h, "error initiating unit reset");
+                flux_future_destroy (f2);
+                f2 = NULL;
+            }
+            proc->f_stop = f2;
+        }
```

Comment:
Since the user can cause failed units I'm not sure this would be useful in real life?

Reply:
ahhh were you thinking a global config?  I was thinking maybe a per-job config in the jobspec.  Like some user will have some really odd setup and we screwed something up and we need to debug it.

Reply:
Oh right, I thought you meant a global config.  Maybe something to consider later if a use case develops?  Or possibly a "grab all properties before reaping" sort of option (since I'm not sure what else you'd get out of preserving a failed unit)?  Anyway, later?

Reply:
yeah that's fine, we can write up an issue afterwards so we don't forget.  I figure "something will go wrong" is basically inevitable :-)

Reply:
>  Or possibly a "grab all properties before reaping" sort of option (since I'm not sure what else you'd get out of preserving a failed unit)? Anyway, later?

I'm just trying to think of when you run `systemctl status` if all of that data is elsewhere if you clean it up.  Like sometimes I think you can see stderr error messages via `systemctl status` if it failed?  But that might be in the journald logs too.  But yeah, can figure it out later if its needed.